aten/src/ATen/native/Sorting.cpp

```diff
@@ -80,20 +83,125 @@ void quick_select_template(
+  TORCH_CHECK(
+      self.device() == q.device(),
+      "quantile() q must be on the same device as the input tensor");
+
+  // Compute output shape: q_size + reduced_size
+  std::vector<int64_t> out_shape;
+  if (_dim && self.dim() > 0) {
+    out_shape = self.sizes().vec();
+    if (keepdim) {
+      out_shape[dim] = 1;
+    } else {
+      out_shape.erase(out_shape.begin() + dim);
+    }
+  } else if (keepdim) {
+    out_shape = std::vector<int64_t>(self.dim(), 1);
+  }
+  if (q.dim() > 0) {
+    out_shape.insert(out_shape.begin(), q.numel());
+  }
+
+  // Check or resize output
+  if (out.numel() > 0) {
+    TORCH_CHECK(
+        out.sizes().vec() == out_shape,
+        "quantile() expected output shape to be ",
+        out_shape,
+        " but got ",
+        out.sizes().vec());
+    TORCH_CHECK(
+        self.scalar_type() == out.scalar_type(),
+        "quantile() out tensor must be same dtype as the input tensor");
+    TORCH_CHECK(
+        self.device() == out.device(),
+        "quantile() out tensor must be on the same device as the input tensor");
+  } else {
+    resize_output(out, out_shape);
+  }
+
+  // Do this check only on cpu to prevent implicit sync on CUDA
+  if (self.device().is_cpu()) {
+    TORCH_CHECK(
+        q.ge(0).logical_and_(q.le(1)).all().item<bool>(),
+        "quantile() q values must be in the range [0, 1]");
+  }
+
+  // If q is scalar, treat as 1D during computations
+  if (q.dim() == 0) {
+    out_shape.insert(out_shape.begin(), q.numel());
+  }
+
+  // Move dimension to reduce as last dimension and flatten if no dim was
+  // specified by the user. Sort to efficiently query kth values.
+  Tensor sorted;
+  if (!_dim) {
+    sorted = std::get<0>(self.flatten().sort());
+  } else if (dim == self.dim() - 1) {
+    sorted = std::get<0>(self.sort());
+  } else {
+    sorted = std::get<0>(self.unsqueeze(-1).transpose_(dim, -1).sort());
+  }
+
+  // View input as reduced_size + reduction_size
+  std::vector<int64_t> in_shape(out_shape.size());
+  std::copy(out_shape.begin() + 1, out_shape.end(), in_shape.begin());
+  in_shape[in_shape.size() - 1] = sorted.size(-1);
+  sorted = sorted.view(in_shape);
+
+  // Ensure converting from int64_t to double won't overflow
+  TORCH_CHECK(
+      sorted.size(-1) <= std::pow(2, 24),
+      "quantile() input tensor is too large");
+  
+  Tensor ranks = q * (sorted.size(-1) - 1);
+  Tensor ranks_below = ranks.toType(kLong);
+  Tensor weights = ranks - ranks_below;
+  Tensor ranks_above = ranks.ceil_().toType(kLong);
+  Tensor values_below = sorted.index_select(-1, ranks_below);
+  Tensor values_above = sorted.index_select(-1, ranks_above);
+
+  // Make out match expected shape for lerp
+  Tensor result = q.dim() == 0
+      ? out.unsqueeze(-1)
+      : out.unsqueeze(-1).transpose_(0, -1).squeeze_(0);
+
+  at::lerp_out(result, values_below, values_above, weights);
+
+  Tensor nan_mask = sorted.isnan().any(-1, false);
```

Comment:
Add a comment explaining the nan-checking here. It's unfortunate that propagating nans correctly here requires three additional kernel launches. It'd be nice if there was a function like `torch.has_nan(tensor) -> bool` that returned true iff the tensor had a nan value. That would at least gate these extra kernels behind a single check. It would also, however, sync the accelerator with the CPU, so it's probably not worthwhile.